BaseClasses.py

```diff
@@ -681,19 +681,108 @@ def can_reach_region(self, spot: str, player: int) -> bool:
+            events_per_player = list(events_per_player_dict.items())
+            del events_per_player_dict
+
+        # If no items logically relevant to a specific world were collected in the previous iteration, then that world's
+        # locations can be skipped in the current iteration. Usually, a world's logic only depends on its own items, but
+        # worlds are allowed to depend on other worlds.
+        # Start with each world only depending on itself.
+        logic_dependent_players: Dict[int, List[int]] = {}
+        worlds = self.multiworld.worlds
+        groups = self.multiworld.groups
+        for player, _locations in events_per_player:
+            if player in groups:
+                # Item link locations have no associated World and use a group ID as the player ID.
+                logic_dependent_players[player] = [player]
+                continue
+            dependencies = worlds[player].player_dependencies
+            if dependencies is not None:
+                for other_player in dependencies:
+                    if other_player in logic_dependent_players:
+                        logic_dependent_players[other_player].append(player)
+                    else:
+                        logic_dependent_players[other_player] = [player]
+
+        # The first iteration must check the locations for all players because it is not known which players might have
+        # reachable locations.
+        players_to_check: Set[int] = set(self.multiworld.regions.location_cache.keys())
+        while players_to_check:
+            received_advancement_players = set()
+            next_events_per_player = []
+
+            for player, locations in events_per_player:
+                if player not in players_to_check:
+                    next_events_per_player.append((player, locations))
+                    continue
+                # Accessibility of all locations is checked first.
+                accessible_locations = []
+                inaccessible_locations = []
+                for location in locations:
+                    # Region accessibility is cached per-player and the cache must be rebuilt when the CollectionState
+                    # for a player changes, but the CollectionState should not change while only checking accessibility,
+                    # so check region accessibility first because, once cached, it should remain cached throughout the
+                    # iteration of `locations`.
+                    # Location.can_reach(), by default, also checks region accessibility if the Location's access rule
+                    # returns True, but because it is known that the region accessibility is cached, it is faster to
+                    # check region accessibility first, even if `location.can_reach(self)` might cause region
+                    # accessibility to be checked a second time.
+                    if location.parent_region.can_reach(self) and location.can_reach(self):
```

Comment:
Reading over this again, I have concerns as to whether it's possible for a location to override `can_reach()` to not check accessibility of its `parent_region` at all and to be placed in an otherwise permanently inaccessible region.

Reply:
An alternative here would be to remove the `location.parent_region.can_reach(self)` and swap `Location.can_reach()` to check region accessibility first because it has pretty much the same result.

Region accessibility caching was added before the comment that checking the Location's access rule computes faster on average, but I'm not sure that comment is correct:
https://github.com/ArchipelagoMW/Archipelago/blob/a05dbac55ffdd1e2f2192421b3fba7fca5341c5c/BaseClasses.py#L1129
The worst case of repeatedly collecting an item and then immediately checking location accessibility would probably be worse because the region accessibility cache has to be repeatedly updated, but is that really a common case such that `location.access_rule` is deemed to be faster?

Reply:
i think the assumption that this is faster is because the latter can trigger a sweep for reachable regions, but the old cache also worked by recomputing the entire thing on miss, whereas the current one will just crash if it misses. it's probably worth checking if that statement is still true or not

Reply:
At least for generating without output and without progression balancing, it is clearly faster to generate on the main branch with the region accessibility checked first, but progression balancing and playthrough calculation would need to be compared too (I'm not sure the change is relevant to patch output).